custom_components/watersmart/client.py

```diff
@@ -121,10 +121,20 @@ async def _authenticate_if_needed(self) -> None:
     async def _authenticate(self) -> None:
         session = self._session
         hostname = self._hostname
+
+        init_response = await session.get(
+            f"https://{hostname}.watersmart.com/index.php/welcome/login"
+        )
+        init_response_text = await init_response.text()
+        init_soup = BeautifulSoup(init_response_text, "html.parser")
+        login_refresh_token_node = init_soup.find("input", {"name":"loginRefreshToken"})
```

Comment:
Would this still work if we moved it [down to here](https://github.com/wbyoung/watersmart/pull/12/files#diff-792bda1744e6797d1797225ee7514382e60170b7294abd4d72de828075575109L136)?

The login flow I'm proposing would be:

- `POST` to try to log in
- If that works, great, no need to do anything else
- Instead of just looking for errors, also search the response page for the `loginRefreshToken`
- if there's a token found, retry the login by doing another `POST`

This will save one more request from being sent for those of us who don't have the token requirement.

Reply:
Hi @wbyoung, thanks , that makes sense.

Let me make the change and test it out.

Reply:
There was no `.error-message`, just the login form being displayed again, so I moved the logic to just above the error checking code (so that we won't have to redo error checking again for the `loginRefreshToken` case). If login is successful, the login form (and thus the `loginRefreshToken`) should not be found on the response page, and all of this logic should be skipped in that case.

Tested on my HA installation and still works. Please take another look and let me know what you think.